### DISCUSSION TITLE
v4.0.0 difference in behavior -- rendering to an initially hidden full-viewport div

### DISCUSSION THREAD

**Discussion by jrochkind** [openseadragon/openseadragon#2279] [General]
When I upgraded to OpenSeadragon v4.0.0 , I have a change in behavior regarding a somewhat odd thing I am doing. I am interested in advice. 

I have isolated a reproduction case from my app, which is a bit weird, but it's a minimal reproduction. 

* In my app, I have a div that is full-screen (position fixed, height/width 100%, top/left 0). It is initially hidden (display: none). 

* I render an OpenSeadragon viewer in it, and put a tilesource in it. 

* Later, as a Javascript action, I show that div. 

You can see a one-page HTML repro of that situation right here with OpenSeadragon 3.1.0. 

https://jrochkind.github.io/osd_v4_diff/osd_v3.1.0_demo.html

Notice that after the hidden div is shown, the image within the OpenSeadragon viewer is exactly fit to screen. That is, pressing the OSD "Go Home" button makes no difference, it's already in fit-to-viewer "home" position. Great!

Now look at the same thing with OpenSeadragon 4.0.0:

https://jrochkind.github.io/osd_v4_diff/osd_v4.0.0_demo.html

Note that the image now starts out zoomed out a bit!  If you press the OSD "Go home" button, a change happens -- pressing the "go home" button changes the zoom level to properly fit the image to the viewer (which is the size of the whole viewport).  I want it the original way with OSD 3.1.0 though!

* The reproduction seems to require the width and height to be 100%, if I make width and height of the initially hidden div some fixed number of `px`, it properly fits to area. 

You can "view source" on these HTML pages to see the fairly simple one-page HTML repro. 

I know I'm doing something maybe a bit weird/unsupported in rendering to an initially hidden div, so I'm not sure if this would actually be considered a bug or not. But it did what I wanted in OSD 3.1.0. 

Is there anything I can do to make it behave the same again in OSD 4.0.0?  I guess I could manually trigger the "go home" action _after_ showing the initially hidden div... but I don't want the animation, I don't really want any delays at all, I really just want the div, when shown, to have the content initially already fit-to-screen as it managed to do in OSD 3.1.0. 

Very curious for feedback or ideas!

**Comment by iangilman** [comment]
Interesting! I'm guessing this has to do with https://github.com/openseadragon/openseadragon/pull/2256 by @pearcetm. That patch improved how we do our resizing for most cases, but in your case, it looks like it hasn't improved things. Yours is an interesting edge case, because when a div is hidden, I believe its width and height is 0. Perhaps because that is technically a square, that explains the change. Sure enough, if I make the window square, it behaves like you want. Note that if you make the window taller than it is wide, the image ends up a little zoomed out from what you want, rather than zoomed in. 

Before that patch, it only took the height into account. That's why your 3.1.0 works when the window is wider than it is tall. If you run the 3.1.0 version on a window that is taller than it is wide (like a tablet for instance), you'll see that it doesn't end up where you'd want it. 

It seems to me we could potentially fix this by adding a special case for if the viewer resizes from 0,0 size. If you want to pursue a fix for that, I can help point you in the right direction, or if you'd just like to file an issue, feel free. 

Otherwise, yes, I think the fix is to tell the viewer to go home when you unhide the div. If you call `viewer.viewport.goHome(true)` it'll do it immediately without any animation.

**Reply by jrochkind** [reply]
Thank you for the quick response!

I don't totally follow, but... it was maybe already broken in cases where the viewport was taller than it was wide? If so, then my implementation was already buggy and I just hadn't noticed, and just adding a `goHome(true)` in the right place is probably the right solution. I will experiment with that, in my alas somewhat convoluted code finding the right place to do it. 

(Even though it's hidden, it has width and height set to 100% of viewport, I was thinking THAT was it's width and height, but maybe not!  I wonder how it ever worked at all then!)

Thank you for your continued awesome stewardship of OpenSeadragon!

**Reply by pearcetm** [reply]
I agree with @iangilman's analysis. This does indeed have to do with viewer resizing from an initially square viewport to a non-square one. 4.0.0 did introduce a change in behavior to take width and height both into account during resize operations, and what you're seeing make a lot of sense based on what size OSD thinks it is while it's hidden. If you look in the debugger, the element's `clientWidth` and `clientHeight` are `0` while hidden. There's an internal method in OSD `_getSafeElemSize` that sets the viewer size to `1, 1` in this case. Since your window has a different aspect ratio, once it's unhidden, the viewer will zoom in/out to compensate for the change in aspect ratio.

I don't think trying to guess exactly what a developer wants to happen after resizing from (1, 1) to the actual size makes sense, especially because `goHome(true)` is an easy way to explicitly define the behavior you want.

**Reply by pearcetm** [reply]
Here's an easy way to do it. Note there's a bit of hackiness needed with `window.setTimeout` - I'll open an issue to discuss this and see if we want to fix it.

This works in the console, because `viewer` is a global variable (window scope) in your demo. You may have to find the right place to put it - like right after you create the viewer - and rename the variable appropriately.

```
viewer.addOnceHandler('resize',function(event){
    window.setTimeout(()=>event.eventSource.viewport.goHome(true));
});
```

This will automatically trigger the `goHome` behavior the first time the viewer is resized, which will happen as soon as it's shown.

**Reply by jrochkind** [reply]
@pearcetm Thank you very much for the code snippet, much appreciated!

It seems to work well. I was having trouble figuring out how to get the `goHome` to happen at the right moment -- not too early -- in my actual app, which is somewhat more complex than this reproduction, and actually involves a Bootstrap 4 'modal'.  

Javascript is not my main language, and I definitely wasn't aware of the very useful OSD-specific utilty `viewer.addOnceHandler` there, or would have thought of hooking into the OSD "resize" event.  So this would have been quite difficult to figure out on my own (and I appreciate the help).  

But I also see your point that it seems hard for OSD to "guess" the right thing to do here, and risks just adding complexity to put in a workaround on top of a workaround. I'm not sure the right solution in OSD and will leave it to y'all -- but if there's a way to detect a resize from "hidden" (if you assume 1x1 is that I guess)  and treat it specially (as an "initial" view that should be set to "home"), my initial reaction is that seems like a potentially good idea.... not sure what happens if you have an OSD that was visible and then you hide and _then_ you show again though, in that case you probably _do_ want to preserve the pan/zoom, not reset it to "home"... I'm not really sure. 

But I think I have a solution to my actual own code (hopefully will hold up to more testing), using your suggestion above. That seems to preserve the behavior I was having in OSD 3.0. (Or maybe even fixes edge-case bugs I had and didn't know about before? Not sure).  Thank you!
